### ISSUE TITLE
Can't focus into text fields, Android + external keyboard

### ISSUE THREAD

**Issue by brucelawson** [facebook/react-native#31820]
In the first Expo example on https://reactnative.dev/docs/textinput - with an external physical keyboard on Android, we don't seem to be able to move the focus into/ between these input fields. It's fine if you tap the screen, and it's fine if Talkback is on and we alt+arrow between the fields. But with Talkback off, and just using the keyboard, we can't get focus into the fields.

## Expected Results
That the fields can be focussed by the keyboard

## Snack, code example, screenshot, or link to a repository:
https://snack.expo.io/PT9eBWga1 (saved from the first example on https://reactnative.dev/docs/textinput)

**Comment by brucelawson** [comment]
Same issue can be seen on Facebook Marketplace. Our workaround was to wrap the textinput in a Pressable  and set focus manually.

**Comment by kriti-s1** [comment]
@brucelawson we are running into a similar issue, can you share the code snippet? Also, are you able to get it working on ios?

**Comment by kriti18singh** [comment]
@brucelawson we are also facing this issue. Could you share the code snippet? Also, are you able to get it working on ios?

**Comment by brucelawson** [comment]
@kriti18singh here's the salient bit of code; the pressable wraps the textinput. We only found the bug affected Android. The Facebook team know about it. Hopefully they'll fix it.

  const setKeyboardFocusToInput = () => {
    if (!focused) {
      focusableRef?.current.focus()
      setFocused(true)
    }
  }

  return (
    <Pressable
      onPress={setKeyboardFocusToInput}
      accessible={isAndroid() ? true : false}
    >

**Comment by brucelawson** [comment]
@kriti18singh I will once we've tested and proved it. (I don't want to let something loose without knowing its side-effects)

**Comment by brucelawson** [comment]
Commit 8 days ago appears related https://github.com/facebook/react-native/commit/d4a498aba2d2843e7a741a31b0c91c6a79a7386c

**Comment by alextait1** [comment]
Hey @brucelawson thanks for submitting this issue, we'll take a closer look soon and report back!

**Comment by alextait1** [comment]
@brucelawson: We've taken a look as a team and will pull this into our prioritized issues, will report back when it's in progress. Thanks again for submitting.

**Comment by pandu-supriyono** [comment]
Hi @alextait1, I would love to know if there are any updates on this matter. In the mean time we're looking at workarounds as described in #30464.

**Comment by alextait1** [comment]
Hey @pandu-supriyono I'm no longer working on this project, maybe @blavalla can provide an update

**Comment by ArturKalach** [comment]
Hello here,
The easiest way it's wrapping TextField in something, I wrote specific lib to support keyboard focus for a11y:
https://www.npmjs.com/package/react-native-a11y

You can wrap `TextInput` into `KeyboardFocusView` and focus `TextInput` on `onFocusChanged` event.

```
    <KeyboardFocusView onFocusChanged={(e) => {
            if(e.nativeEvent.isFocused) {
              ref.focus()
            }
          }}>
            <TextInput
              ref={inputRef}
            />
          </KeyboardFocusView>
```
I can imagine more elegant solution on old arch, but new arch is hell challenging

**Comment by ArturKalach** [comment]
There is elegent solution for `old arch`.
You can override native `EditText`:
1.  Create `RCTEditText` component on android and extend `ReactEditText`
```

import android.content.Context;
import android.graphics.Rect;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;

import com.facebook.react.views.textinput.ReactEditText;

public class RCTEditText extends ReactEditText {
    static final int ENTER_KEY_CODE = KeyEvent.KEYCODE_ENTER;
    static final int MULTILINE_TYPE = EditorInfo.TYPE_TEXT_FLAG_MULTI_LINE;

    public RCTEditText(Context context) {
        super(context);
    }

    private boolean isKeyHandled(int keyCode) {
        boolean isMultiline = (this.getInputType() & MULTILINE_TYPE) == 0;
        return keyCode != ENTER_KEY_CODE
                || !isMultiline
                || !this.getBlurOnSubmit();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (this.isKeyHandled(keyCode)) {
            return super.onKeyDown(keyCode, event);
        }
        return false;
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        if (this.isKeyHandled(keyCode)) {
            return super.onKeyUp(keyCode, event);
        }
        return false;
    }

    @Override
    public boolean requestFocus(int direction, Rect previouslyFocusedRect) {
        if (direction == View.FOCUS_FORWARD || direction == View.FOCUS_BACKWARD) {
            this.requestFocusFromJS();
        }
        return super.requestFocus(direction, previouslyFocusedRect);
    }
}

```

2. Create `RCTEditTextManager` and extend `ReactTextInputManager`
```

import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.views.textinput.ReactEditText;
import com.facebook.react.views.textinput.ReactTextInputManager;

class RCTEditTextManager extends ReactTextInputManager {
    // Copied react native text input manager, just copy pasted
    @Override
    public ReactEditText createViewInstance(ThemedReactContext context) {
        //Changed this line
        ReactEditText editText = new RCTEditText(context);
        int inputType = editText.getInputType();
        editText.setInputType(inputType & -131073);
        editText.setReturnKeyType("done");
        return editText;
    }
}
```

3. Create `RCTEditTextPackage`:

```

import androidx.annotation.NonNull;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class RCTEditTextPackage implements ReactPackage {
    @NonNull
    @Override
    public List<NativeModule> createNativeModules(@NonNull ReactApplicationContext reactContext) {
        return Collections.emptyList();
    }

    @NonNull
    @Override
    public List<ViewManager> createViewManagers(@NonNull ReactApplicationContext reactContext) {
        List<ViewManager> viewManagers = new ArrayList<>();
        viewManagers.add(new RCTEditTextManager());
        return viewManagers;
    }
}

```

4. Add package in `MainApplication`
```
     List<ReactPackage> packages = new PackageList(this).getPackages();
      packages.add(new RCTEditTextPackage());
```

Example of files in derictory:
<img width="292" alt="image" src="https://user-images.githubusercontent.com/112642862/235763921-c00d7edd-3ba4-41ec-8194-c83ff5b9c5da.png">

Example of `getPackages()`
<img width="707" alt="image" src="https://user-images.githubusercontent.com/112642862/235763998-769208d7-f2fb-488b-a15a-fb27e9ca5a07.png">

I am not sure whether it will work on `new arch`, but it works for me

**Comment by nedgip** [comment]
Are there any plans to fix the keyboard accessibility of the  `<TextInput>`?

The workaround https://github.com/facebook/react-native/issues/30464 to wrap the `<TextInput>` in a `<Pressable>` doesn't work with the Tab key and doesn't support keyboard users who also use a screen reader.

The React Native `<TextInput>` should be accessible using a physical keyboard out-of-the-box without implementing workarounds as it is a critical element for many apps.

**Comment by sharad-incapsulate** [comment]
Is there any proper solution to make it work as expected ?

**Comment by chimiWangchukWangdi** [comment]
Same here

**Comment by brucelawson** [comment]
Great to see the priority given to this accessibility bug! It's 2 years, 8 months and 29 days since I posted it.

**Comment by rabbitmouse** [comment]
same here, anyone known how to fix it?

**Comment by ArturKalach** [comment]
Hello there,

After some work with the native and a keyboard focus, I finally found a workaround for TextInput focusability. You can find it in the [react-native-external-keyboard](https://www.npmjs.com/package/react-native-external-keyboard) package, in the component called KeyboardExtendedTextInput.


The solution works for Android and iOS, and support new arch (bridgeless is not supported).

**Comment by proohit** [comment]
Is there any news on this? We also need to make sure TextInput is focusable with an external keyboard.

**Comment by BogdanCerovac** [comment]
Can not re-open this, so I created a new one: https://github.com/facebook/react-native/issues/45801 (with additional related test case)...

**Comment by BogdanCerovac** [comment]
@jamonholmgren - thank you for re-opening this.

Temporary solution seems to be using the packages made by @ArturKalach, but I hope React Native can support external keyboard out of the box for all the folks not aware of the issue.

**Comment by louie-succio** [comment]
Any update on if this will be fixed?  Our team is having the same issue as we try to complete our Accessibility audits

**Comment by timoisalive** [comment]
In addition to the issue of not being able to focus on `TextInput` with keyboard, we are facing an issue of not being able to make `TextInput` NON-focusable with the screen reader!

In some scenarios we'd like the screen reader to not focus on the `TextInput` at all, but `focusable={false}` nor `accessible={false}` does the trick. On iOS we don't have this issue.

**Comment by Sheiicker** [comment]
I worked on a workaround that allows us to tab between inputs.
Here is how i made it:

- Wrap `TextInput` in a homemade `CustomInput` with something like `Pressable` or `TouchableHighlight`. (I used the `TouchableHighlight` from react-native-gesture-handler lib)
- You can now set your own method `focus` in `CustomInput` which will switch on/off the prop `editable` of your `TextInput` instance. You also should add a reference focusing of your `TextInput` with a callback to focus it.

Note that when the prop `editable` is set to `false` in your `TextInput`, you can press tab on each TouchableHighlight and go to the next element.

**Comment by d1fferenz** [comment]
To be compliant to the EAA (EU Accessibility Act) we need support on these kind of issues! Like the focus-bug described in this issue, keyboard traps in general and also to be able to turn of the focusability of no-interactive elements. On iOS most of the things are working out-of-the-box, but android behaves very special.

**Comment by shivsuthanmhub** [comment]
Yes the issue got fixed with React-native version upgrade of 0.80.0 or higher.

FYI

https://github.com/user-attachments/assets/46059fc1-125b-4c2c-8291-d3a3dd16b55a

**Comment by BogdanCerovac** [comment]
Had a similar/cloned issue - verified with 0.83.1 and it is (still) working; https://github.com/facebook/react-native/issues/45801#issuecomment-3700587513

**Comment by popsiclelmlm** [comment]
Triage note: this looks like it may now be fixed in current React Native releases.

The relevant fix appears to be 3420eb87b01f42092c1691a1f6544f20d244cc60 / #48547, which changed Android TextInput focus handling so native `requestFocus` calls are no longer swallowed. The changelog for that commit says: "TextInputs can now receive focus via external keyboard".

That also matches the follow-up reports here and in the cloned issue #45801:
- https://github.com/facebook/react-native/issues/31820#issuecomment-3486626346 reports this fixed after upgrading to RN 0.80.0 or higher.
- https://github.com/facebook/react-native/issues/45801#issuecomment-3700587513 verifies physical keyboard focus on TextInput with RN 0.83.1.

I have not re-tested this locally myself, but based on the landed fix and the latest confirmations, this issue may be ready to close as fixed unless someone can still reproduce it on a current RN release with a minimal repro.
